### ISSUE TITLE
FR:add chinese input method

### ISSUE THREAD

**Issue by seacatwu** [koreader/koreader#1938]
When we use search function, it needs chinese IM........
can we use kindle native IM?

**Comment by houqp** [comment]
It might not be trivia to integrate with kindle's IM. Probably easier to just implement our own im module.

**Comment by seacatwu** [comment]
great

**Comment by frankyifei** [comment]
I think it needs great effort to implement.

**Comment by houqp** [comment]
yes, either way, it will be a non-triva work :)

**Comment by Hzj-jie** [comment]
I am looking at VirtualKeyboard.lua. It looks pretty hard to extend, which is the first issue we are facing to support input method other than English and Russian. If you have no opinion, I will work on it.
Considering it would be a very large change, I will first prepare some design doc.
Please kindly let me know if you agree on this plan.

**Comment by houqp** [comment]
@Hzj-jie go ahead if you want to take a stab at it :)

**Comment by seacatwu** [comment]
just waiting ：）

**Comment by Hzj-jie** [comment]
Sorry for the late response. I was busy with my bread work recently. But now it's time for me to fulfill my promise.

I spent some time to investigate the possibility, and realized KOReader is becoming a huge system with too many functions in one binary. And a tiny error can eventually break the entire system. From my own opinion, this is not good. I prefer the way Unix works - the combination of small and focus tools. So instead of adding a input method - as this by mentioned, instead of inputting Chinese, searching books with Chinese names is more important - I would suggest to use an index file. Then we can use background tasks to index file names or even file contents without changing KOReader core logic.
And besides indexing files, I would like to add another background task to export book marks and comments to a single file, as Kindle does.

Please kindly let me know your opinions.

**Comment by houqp** [comment]
> Sorry for the late response. I was busy with my bread work recently. But now it's time for me to fulfill my promise.

Good to see you back :)

> I spent some time to investigate the possibility, and realized KOReader is becoming a huge system with too many functions in one binary. And a tiny error can eventually break the entire system. From my own opinion, this is not good. I prefer the way Unix works - the combination of small and focus tools. 

Can't agree more. I have been thinking of splitting out the UI and rendering code apart for awhile. It will be a big win for KOReader. This way, UI crash will not crash the whole reader and the UI loop will not be blocked by the rendering job. It also makes it possible to plug in other UI frontends written in any language. A UI frontend can even connect to the headless reader core remotely through network.

> So instead of adding a input method - as this by mentioned, instead of inputting Chinese, searching books with Chinese names is more important - I would suggest to use an index file. Then we can use background tasks to index file names or even file contents without changing KOReader core logic.

Hm.. I think indexing and input method are two separate issues. If that's what you want to explore, we should move the discussion to a separate issue. You should also join effort with @chrox on the index design since the new KOReader home might need to introduce indexes too.

> And besides indexing files, I would like to add another background task to export book marks and comments to a single file, as Kindle does.

We should also evaluate various solutions for setting up background task in Lua. Lua does not have good builtin support for setting up real parallel background tasks.

**Comment by Hzj-jie** [comment]
Is there any issue to track index change?
My initial plan is to use bash scripts, we can use various existing tools, grep, awk,cut, etc, without writting complex code.

**Comment by houqp** [comment]
Let's move the discussion to https://github.com/koreader/koreader/issues/2122

**Comment by seacatwu** [comment]
thank you Hzj-jie.
That's ok.

**Comment by Jason-summer** [comment]
It has been so long since the last discussion, and it's sad and true that we need Chinese input (and maybe other Asian language input methods). Is there anyone who have idea how to solve it.

And I see there is a Japanese keyboard layout, but it seen that doesn't work.

**Comment by Frenzie** [comment]
@Jason-summer How do you mean doesn't work?

**Comment by Jason-summer** [comment]
@Frenzie Nothing. Sorry, I have made a mistake. I supposed it's click the global to switch the language instead of long-press. Sorry.

By the way. I found that Korean keyboard layout use a special way to form a individual character, just like a way of input of Chinese character called wubi, so maybe we still have some ways to try.

**Comment by Bacuer** [comment]
Very need Chinese input method

**Comment by vietchovui** [comment]
I need Vietnamese input method. Hope this will be available soon.

**Comment by finelog** [comment]
hello, I've been using koreader for almost two years, yet just found out koreader does not support chinese input recently. and I am a Chinese in China! 🤣 

anyway, my first thought was why not using [librime](https://github.com/rime/librime).
so I was trying this idea, and having some progress, thought I should share to you all.
aware though, it's very primary, still a long road to go. :)

![SCR-20220730-wj1](https://user-images.githubusercontent.com/9196730/181923811-b4b708f5-7db6-42d9-8955-f643f4436f1b.png)
the screenshot above is lua code successfully using librime input functionality
the next steps will be:

- [ ] a chinese input keyboard
- [ ] connect keyboard with lua and librime

but, it won't be very fast, beause
1. I am not very familiar with lua frontend stuff, take time to get comfort with it.
2. I won't be having free time until October

so, if it's all by me, it probably won't done until this year's ending.

I attach my code here, so if anyone is interest, please feel free to use it.

[rime-lua.zip](https://github.com/koreader/koreader/files/9226816/rime-lua.zip)

**Comment by ShellWen** [comment]
> hello, I've been using koreader for almost two years, yet just found out koreader does not support chinese input recently. and I am a Chinese in China! 🤣
> 
> anyway, my first thought was why not using [librime](https://github.com/rime/librime). so I was trying this idea, and having some progress, thought I should share to you all. aware though, it's very primary, still a long road to go. :)
> 
> ![SCR-20220730-wj1](https://user-images.githubusercontent.com/9196730/181923811-b4b708f5-7db6-42d9-8955-f643f4436f1b.png) the screenshot above is lua code successfully using librime input functionality the next steps will be:
> 
>     * [ ]  a chinese input keyboard
> 
>     * [ ]  connect keyboard with lua and librime
> 
> 
> but, it won't be very fast, beause
> 
>     1. I am not very familiar with lua frontend stuff, take time to get comfort with it.
> 
>     2. I won't be having free time until October
> 
> 
> so, if it's all by me, it probably won't done until this year's ending.
> 
> I attach my code here, so if anyone is interest, please feel free to use it.
> 
> [rime-lua.zip](https://github.com/koreader/koreader/files/9226816/rime-lua.zip)

Nice job! I wanna take part in, but im not professional at lua. I can wait for it.
